.github/workflows/test-linux.yml

```diff
@@ -0,0 +1,172 @@
+          # Build and install both qiskit and qiskit-terra so that any optionals
+          # depending on `qiskit` will resolve correctly.
+      - name: Install Optional packages
+        run: |
+          set -e
+          source test-job/bin/activate
+          python -m pip install -r requirements-optional.txt -c constraints.txt
+          python -m pip check
+        if: ${{ inputs.install-optionals }}
+      - name: Install optional non-Python dependencies
+        run: |
+          set -e
+          sudo apt-get update
+          sudo apt-get install -y graphviz
+        if: ${{ inputs.install-optionals }}
+      - name: Run Rust Tests
+        if: inputs.test-rust == true
+        # Note that we use the virtual env with Qiskit installed to run the Rust
+        # tests since some of them still depend on Qiskit's Python API via PyO3.
+        run: |
+          source test-job/bin/activate
+          python tools/report_numpy_state.py
+          python tools/run_cargo_test.py
+      - name: Run Python tests
+        run: |
+          set -e
+          source test-job/bin/activate
+          python tools/report_numpy_state.py
+          mkdir -p /tmp/terra-tests
+          cp -r test /tmp/terra-tests/.
+          cp .stestr.conf /tmp/terra-tests/.
+          cp -r .stestr /tmp/terra-tests/. || :
+          pushd /tmp/terra-tests
+          export PYTHONHASHSEED=$(python -S -c "import random; print(random.randint(1, 4294967295))")
+          echo "PYTHONHASHSEED=$PYTHONHASHSEED"
+          stestr run
+          popd
+        env:
+          QISKIT_PARALLEL: FALSE
+          QISKIT_IGNORE_USER_SETTINGS: TRUE
+          RUST_BACKTRACE: 1
+      - name: Install Image dependencies
+        run: |
+          set -e
+          virtualenv image_tests
+          image_tests/bin/python -m pip install -U \
+            -c constraints.txt \
+            -r requirements.txt \
+            -r requirements-dev.txt \
+            -r requirements-optional.txt \
+            -e .
+          sudo apt-get update
+          sudo apt-get install -y graphviz pandoc
+          image_tests/bin/pip check
+        if: ${{ inputs.test-images }}
+      - name: Run image tests
+        run: |
+          echo "##vso[task.setvariable variable=HAVE_VISUAL_TESTS_RUN;]true"
+          image_tests/bin/python -m unittest discover -v test/visual
+        if: ${{ inputs.test-images }}
+        env:
+          # Needed to suppress a warning in jupyter-core 5.x by eagerly migrating to
+          # a new internal interface that will be the default in jupyter-core 6.x.
+          # This variable should become redundant on release of jupyter-core 6.
+          JUPYTER_PLATFORM_DIRS: 1
+      - name: Store image tests diff
+        uses: actions/upload-artifact@v4
+        if: ${{ inputs.test-images && failure()}}
+        with:
+          name: image-test-failure-img-diffs
+          path: |
+            ./test/visual/mpl/graph/graph_results
+          if-no-files-found: error
```

Comment:
We should have enough job parallelism on linux to be able to split these  out into separate jobs instead of serially executed tasks. This would reduce the runtime on this job which will be the critical path for the preliminary stage.